README.md
# note-parser [![npm](https://img.shields.io/npm/v/note-parser.svg)](https://www.npmjs.com/package/note-parser)

[![Build Status](https://travis-ci.org/danigb/note-parser.svg?branch=master)](https://travis-ci.org/danigb/note-parser) [![Code Climate](https://codeclimate.com/github/danigb/note-parser/badges/gpa.svg)](https://codeclimate.com/github/danigb/note-parser)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](https://github.com/feross/standard)

Parse note names (in [scientific notation](https://en.wikipedia.org/wiki/Scientific_pitch_notation)) with javascript. Given a string, obtain a hash
with note properties (including midi number and frequency)

If you need parse interval names take a look to [interval-notation](https://github.com/danigb/interval-notation)

## Usage

Install via npm: `npm i --save note-parser` and require it:

```js
var parser = require('note-parser')
parser.parse('c#4') // => { letter: 'C', acc: '#', ... midi: 61, freq: 277.1826309768721 }
```

The returned object will contain:

- `letter`: the uppercase letter of the note
- `acc`: the accidentals of the note (only sharps or flats)
- `pc`: the pitch class (letter + acc)
- `step`: s a numeric representation of the letter. It's an integer from 0 to 6 where 0 = C, 1 = D ... 6 = B
- `alt`: a numeric representation of the accidentals. 0 means no alteration,
positive numbers are for sharps and negative for flats
- `chroma`: a numeric representation of the pitch class. It's like midi for
pitch classes. 0 = C, 1 = C#, 2 = D ... It can have negative values: -1 = Cb.

If the note name has octave, the returned object will additionally have:

- `oct`: the octave number (as integer)
- `midi`: the midi number
- `freq`: the frequency (using tuning parameter as base)

If the parameter `isTonic` is set to true another property is included:

- `tonicOf`: the rest of the string that follows note name (left and right trimmed)  

#### Midi note number and frequency

If you are interested only in midi numbers or frequencies, you can use `midi` function:

```js
var midi = require('note-parser').midi
midi('A4') // => 69
midi('blah') // => null
```

or the `freq` function:

```js
var freq = require('note-parser').freq
freq('A4') // => 440
```

You can use also the `midiToFreq` function if you need:

```js
var midiToFreq = require('note-parser').midiToFreq
midiToFreq(69) // => 440
midiToFreq(69, 444) // => 444
```

### Build the string back

With the `build` function you can convert back to string:

```js
parser.build(parser.parse('cb2')) // => 'Cb2'
```

The `build` function also accepts `step, alteration, octave` parameters:

```js
parser.build(3, -2, 4) // => 'Fbb4'
```

## Tests and documentation

You can read the [generated API documentation here](https://github.com/danigb/note-parser/blob/master/API.md)

To run the test clone this repo and:

```
npm i
npm test
```

## License

MIT License


CHANGELOG
[MASTER]

[1.2.0]
- Add build function

[1.1.0]
- Add toFreq function

[1.0.0]
- Add better REGEX support
- Add function for each property
- Add chroma, alt

[0.9.0]
- Add name and remove toString

[0.8.0]
- Add plus and minus notation (impro-visor style)
- Move to standard js style

[0.7.0]
- Add midi note
- Add frequency

[0.6.0]
- Add defaultOctave param
- Add defaultValue param

[0.5.0]
- Octave is, by default, 4
- Rename pitchClass to pc
- Rename octave to oct
- Rename accidentals acc

[0.4.0]
- Add parse.toString method

[0.3.0]
- Rename to note-parser
- Fail fast: parser error now throws exception

[0.2.0]
- Parse a parsed note return itself

[0.1.0]
- First version
- Parse notes with accidentals


test/freq.js
'use strict'
var vows = require('vows')
var assert = require('assert')

var parse = require('../')

vows.describe('note freq').addBatch({
  'middle A is 440': function () {
    assert.equal(parse('A', 4).freq, 440)
  },
  'octaves': function () {
    assert.equal(parse('a2').freq, 110)
    assert.equal(parse('a3').freq, 220)
    assert.equal(parse('a4').freq, 440)
    assert.equal(parse('a5').freq, 880) // theorical limit
  },
  'accidentals': function () {
    assert.equal(parse('C').freq, 261.6255653005986)
    assert.equal(parse('C#').freq, parse('Db').freq)
    assert.equal(parse('C##').freq, parse('D').freq)
    assert.equal(parse('Cb4').freq, parse('B3').freq)
    assert.equal(parse('Cbb4').freq, parse('Bb3').freq)
  }

}).export(module)


test/midi.js
var vows = require('vows')
var assert = require('assert')

var parse = require('../')

vows.describe('note midi').addBatch({
  'middle C is 60': function () {
    // Wikipedia: Middle C is designated C4 in scientific pitch notation
    assert.equal(parse('c', 4).midi, 60)
  },
  'octaves': function () {
    assert.equal(parse('c-2').midi, -12)
    assert.equal(parse('c-1').midi, 0)
    assert.equal(parse('c0').midi, 12)
    assert.equal(parse('c1').midi, 24)
    assert.equal(parse('a4').midi, 69)
    assert.equal(parse('g9').midi, 127) // theorical limit
  },
  'accidentals': function () {
    assert.equal(parse('C').midi, 60)
    assert.equal(parse('C#').midi, 61)
    assert.equal(parse('C##').midi, 62)
    assert.equal(parse('Cb').midi, 59)
    assert.equal(parse('Cbb').midi, 58)
  }

}).export(module)


test/name.js
var vows = require('vows')
var assert = require('assert')

var parse = require('..')

vows.describe('Note name').addBatch({
  'note name': function () {
    assert.equal(parse('C#3').name, 'c#3')
    assert.equal(parse('C').name, 'c4')
    assert.equal(parse('db--').name, 'db2')
  }
}).export(module)


index.js
'use strict'

// util
function fillStr (s, num) { return Array(num + 1).join(s) }
function isNum (x) { return typeof x === 'number' }

var REGEX = /^([a-gA-G])(#{1,}|b{1,}|x{1,}|)(-?\d*)\s*(.*)\s*$/
/**
 * A regex for matching note strings in scientific notation.
 *
 * @name regex
 * @function
 * @return {RegExp} the regexp used to parse the note name
 *
 * The note string should have the form `letter[accidentals][octave][element]`
 * where:
 *
 * - letter: (Required) is a letter from A to G either upper or lower case
 * - accidentals: (Optional) can be one or more `b` (flats), `#` (sharps) or `x` (double sharps).
 * They can NOT be mixed.
 * - octave: (Optional) a positive or negative integer
 * - element: (Optional) additionally anything after the duration is considered to
 * be the element name (for example: 'C2 dorian')
 *
 * The executed regex contains (by array index):
 *
 * - 0: the complete string
 * - 1: the note letter
 * - 2: the optional accidentals
 * - 3: the optional octave
 * - 4: the rest of the string (trimmed)
 *
 * @example
 * var parser = require('note-parser')
 * parser.regex.exec('c#4')
 * // => ['c#4', 'c', '#', '4', '']
 * parser.regex.exec('c#4 major')
 * // => ['c#4major', 'c', '#', '4', 'major']
 * parser.regex().exec('CMaj7')
 * // => ['CMaj7', 'C', '', '', 'Maj7']
 */
function regex () { return REGEX }

var SEMITONES = [0, 2, 4, 5, 7, 9, 11]
/**
 * Parse a note name in scientific notation an return it's components,
 * and some numeric properties including midi number and frequency.
 *
 * @name parse
 * @function
 * @param {String} note - the note string to be parsed
 * @param {Boolean} isTonic - true if the note is the tonic of something.
 * If true, en extra tonicOf property is returned. It's false by default.
 * @param {Float} tunning - The frequency of A4 note to calculate frequencies.
 * By default it 440.
 * @return {Object} the parsed note name or null if not a valid note
 *
 * The parsed note name object will ALWAYS contains:
 * - letter: the uppercase letter of the note
 * - acc: the accidentals of the note (only sharps or flats)
 * - pc: the pitch class (letter + acc)
 * - step: s a numeric representation of the letter. It's an integer from 0 to 6
 * where 0 = C, 1 = D ... 6 = B
 * - alt: a numeric representation of the accidentals. 0 means no alteration,
 * positive numbers are for sharps and negative for flats
 * - chroma: a numeric representation of the pitch class. It's like midi for
 * pitch classes. 0 = C, 1 = C#, 2 = D ... It can have negative values: -1 = Cb.
 * Can detect pitch class enhramonics.
 *
 * If the note has octave, the parser object will contain:
 * - oct: the octave number (as integer)
 * - midi: the midi number
 * - freq: the frequency (using tuning parameter as base)
 *
 * If the parameter `isTonic` is set to true, the parsed object will contain:
 * - tonicOf: the rest of the string that follows note name (left and right trimmed)
 *
 * @example
 * var parse = require('note-parser').parse
 * parse('Cb4')
 * // => { letter: 'C', acc: 'b', pc: 'Cb', step: 0, alt: -1, chroma: -1,
 *         oct: 4, midi: 59, freq: 246.94165062806206 }
 * // if no octave, no midi, no freq
 * parse('fx')
 * // => { letter: 'F', acc: '##', pc: 'F##', step: 3, alt: 2, chroma: 7 })
 */
function parse (str, isTonic, tuning) {
  if (typeof str !== 'string') return null
  var m = REGEX.exec(str)
  if (!m || !isTonic && m[4]) return null

  var p = { letter: m[1].toUpperCase(), acc: m[2].replace(/x/g, '##') }
  p.pc = p.letter + p.acc
  p.step = (p.letter.charCodeAt(0) + 3) % 7
  p.alt = p.acc[0] === 'b' ? -p.acc.length : p.acc.length
  p.chroma = SEMITONES[p.step] + p.alt
  if (m[3]) {
    p.oct = +m[3]
    p.midi = p.chroma + 12 * (p.oct + 1)
    p.freq = midiToFreq(p.midi, tuning)
  }
  if (isTonic) p.tonicOf = m[4]
  return p
}

var LETTERS = 'CDEFGAB'
function acc (n) { return !isNum(n) ? '' : n < 0 ? fillStr('b', -n) : fillStr('#', n) }
function oct (n) { return !isNum(n) ? '' : '' + n }

/**
 * Create a string from a parsed object or `step, alteration, octave` parameters
 * @param {Object} obj - the parsed data object
 * @return {String} a note string or null if not valid parameters
 * @since 1.2
 * @example
 * parser.build(parser.parse('cb2')) // => 'Cb2'
 * parser.build(3, -1, 4) // => 'Fb4'
 */
function build (s, a, o) {
  if (s === null || typeof s === 'undefined') return null
  if (s.step) return build(s.step, s.alt, s.oct)
  if (s < 0 || s > 6) return null
  return LETTERS.charAt(s) + acc(a) + oct(o)
}

/**
 * Given a midi number, return its frequency
 * @param {Integer} midi - midi note number
 * @param {Float} tuning - (Optional) the A4 tuning (440Hz by default)
 * @return {Float} frequency in hertzs
 */
function midiToFreq (midi, tuning) {
  return Math.pow(2, (midi - 69) / 12) * (tuning || 440)
}

/**
 * Get midi of a note
 *
 * @name midi
 * @function
 * @param {String|Integer} note - the note name or midi number
 * @return {Integer} the midi number of the note or null if not a valid note
 * or the note does NOT contains octave
 * @example
 * var parser = require('note-parser')
 * parser.midi('A4') // => 69
 * parser.midi('A') // => null
 */
function midi (note) {
  var p = parse(note)
  return p ? p.midi : null
}

var parser = { parse: parse, build: build, regex: regex, midi: midi, midiToFreq: midiToFreq }
var FNS = ['letter', 'acc', 'pc', 'step', 'alt', 'chroma', 'oct', 'freq']
FNS.forEach(function (name) {
  parser[name] = function (src) {
    var p = parse(src)
    return p && (typeof p[name] !== 'undefined') ? p[name] : null
  }
})

/**
 * Get freq of a note in hertzs (in a well tempered 440Hz A4)
 *
 * @name freq
 * @function
 * @param {String} note - the note name
 * @return {Float} the freq of the number if hertzs or null if not valid note
 * or the note does NOT contains octave
 * @example
 * var parser = require('note-parser')
 * parser.freq('A4') // => 440
 * parser.freq('A') // => null
 */

module.exports = parser
